### PR TITLE
Handle TLS 1.3 client certification auth failure

### PR DESCRIPTION
**Pull request by andreachild**
Fix NullPointerException being thrown by outbound encoder due to null remote host which can happen with TLS 1.3 client certification auth failure. Client cert auth failure can happen after a successful handshake, causing an inbound SSLException. Changed the inbound handler to detect and store the SSLException in the channel attributes and changed the outbound encoder to detect if remote address is not available and there was an SSLException. Changed the integration test to log the stack trace of the caught exception for easier troubleshooting as these SSL tests have been flaky.

Tested the fix by running GremlinServerSslIntegrateTest and gremlin-console remote connection to gremlin-server with ssl enabled and client auth required.

### CODE REVIEW

**Comment by xiazcy** [comment]
VOTE +1

**Comment by Cole-Greer** [comment]
Thanks, it's great to know why that test has been so troublesome and to finally get it pinned down.

VOTE +1

File: gremlin-driver/src/main/java/org/apache/tinkerpop/gremlin/driver/handler/HttpGremlinRequestEncoder.java [hunk 1]
```diff
@@ -68,14 +71,15 @@ protected void encode(final ChannelHandlerContext channelHandlerContext, final R
                     requestMessage));
         }
 
+        InetSocketAddress remoteAddress = getRemoteAddress(channelHandlerContext.channel());
```
Comment (kenhuuu) [inline gremlin-driver/src/main/java/org/apache/tinkerpop/gremlin/driver/handler/HttpGremlinRequestEncoder.java:74]: Nit: add `final`. Please check over other parts that may need similar update.

File: gremlin-driver/src/main/java/org/apache/tinkerpop/gremlin/driver/handler/GremlinResponseHandler.java [hunk 2]
```diff
@@ -44,6 +46,7 @@
 public class GremlinResponseHandler extends SimpleChannelInboundHandler<ResponseMessage> {
     private static final Logger logger = LoggerFactory.getLogger(GremlinResponseHandler.class);
     private static final AttributeKey<ResponseException> CAUGHT_EXCEPTION = AttributeKey.valueOf("caughtException");
+    private static final AttributeKey<Throwable> SSL_EXCEPTION = AttributeKey.valueOf("sslException");
```
Comment (kenhuuu) [inline gremlin-driver/src/main/java/org/apache/tinkerpop/gremlin/driver/handler/GremlinResponseHandler.java:49]: This is duplicated in two classes. Consider moving to `StateKey`.

**Comment by kenhuuu** [comment]
VOTE +1, pending fixes to minor nits.

File: gremlin-driver/src/main/java/org/apache/tinkerpop/gremlin/driver/handler/HttpGremlinRequestEncoder.java (hunk 1 above)
Comment (andreachild) [reply gremlin-driver/src/main/java/org/apache/tinkerpop/gremlin/driver/handler/HttpGremlinRequestEncoder.java]: Added a couple finals

File: gremlin-driver/src/main/java/org/apache/tinkerpop/gremlin/driver/handler/GremlinResponseHandler.java (hunk 2 above)
Comment (andreachild) [reply gremlin-driver/src/main/java/org/apache/tinkerpop/gremlin/driver/handler/GremlinResponseHandler.java]: Removed duplication by changing the AttributeKey in the GremlinResponseHandler to public (as it is the class that adds the attribute to the context) and changing HttpGremlinRequestEncoder to reuse the AttributeKey from GremlinResponseHandler.

**Pull request merged**